### ISSUE TITLE
Issue with Nginx: 502 Bad Gateway

### ISSUE THREAD

**Issue by Ikfes** [DragonMinded/bemaniutils#16]
First off, this is an amazing project for a small tinkerer like me to try out at my home during covid times.. Thank you so much for making this in the first place.

I have a problem at final stages though and would appreciate any help with it.

I have installed all the required packages, and all stuff in the requirements.txt including wheel and uwsgi via pip3.

I have tested everything via development tools so far and verified, that it is working as intended. The database is also populated with charts from single game I will use it with. It is also verified to work with traffic gen.

Then I managed to get the wsgi files running with the uwsgi command.

```
root@server~# sudo ps wwaux | grep -i uwsgi
root     26928  0.4  0.7 265756 58168 pts/4    Sl+  17:54   0:00 uwsgi --ini /root/examples/uwsgi/api.ini
root     26933  0.0  0.6 265756 48304 pts/4    Sl+  17:54   0:00 uwsgi --ini /root/examples/uwsgi/api.ini
root     26934  0.0  0.6 265756 48304 pts/4    Sl+  17:54   0:00 uwsgi --ini /root/examples/uwsgi/api.ini
root     26937  0.0  0.6 265756 48304 pts/4    Sl+  17:54   0:00 uwsgi --ini /root/examples/uwsgi/api.ini
root     26940  0.4  0.7 268624 60976 pts/5    Sl+  17:54   0:01 uwsgi --ini /root/examples/uwsgi/services.ini
root     26945  0.0  0.6 268624 50928 pts/5    Sl+  17:54   0:00 uwsgi --ini /root/examples/uwsgi/services.ini
root     26946  0.0  0.6 268624 50928 pts/5    Sl+  17:54   0:00 uwsgi --ini /root/examples/uwsgi/services.ini
root     26949  0.0  0.6 268624 50928 pts/5    Sl+  17:54   0:00 uwsgi --ini /root/examples/uwsgi/services.ini
root     26952  0.6  0.8 271476 64104 pts/6    Sl+  17:54   0:01 uwsgi --ini /root/examples/uwsgi/frontend.ini
root     26957  0.0  0.6 271476 54220 pts/6    Sl+  17:54   0:00 uwsgi --ini /root/examples/uwsgi/frontend.ini
root     26958  0.0  0.6 271476 54220 pts/6    Sl+  17:54   0:00 uwsgi --ini /root/examples/uwsgi/frontend.ini
root     26961  0.0  0.6 271476 54220 pts/6    Sl+  17:54   0:00 uwsgi --ini /root/examples/uwsgi/frontend.ini
root     26999  0.0  0.0  13144  1016 pts/1    S+   17:58   0:00 grep --color=auto -i uwsgi

```

Next I configured nginx:
```

server {
	listen 80 default_server;
	listen [::]:80 default_server;

	root /var/www/html;

	index index.html index.htm index.nginx-debian.html;

	server_name _;

	location / {
		# First attempt to serve request as file, then
		# as directory, then fall back to displaying a 404.
		try_files $uri $uri/ =404;
	}

}

server {
    listen 573;
    server_tokens off;

    location / {
        include uwsgi_params;
        uwsgi_pass unix:/mnt/sock/services.sock;
    }
}

server {
    server_name **<mydomain>**;
    listen 8573 ssl;
    server_tokens off;

    gzip on;
    gzip_types text/html text/css text/plain application/javascript application/xml application/json;
    gzip_min_length 1000;

    ssl_certificate /etc/letsencrypt/live/**<mydomain>**/fullchain.pem;
    ssl_certificate_key /etc/letsencrypt/live/**<mydomain>**/privkey.pem;

    location / {
        include uwsgi_params;
        uwsgi_pass unix:/mnt/sock/frontend.sock;
    }

    location ^~ /static/ {
        include  /etc/nginx/mime.types;
        root /root/virtualenv/bemani/wsgi/bemani/frontend;
    }
}

server {
    server_name **<mydomain>**;
    listen 9573 ssl;
    server_tokens off;

    gzip on;
    gzip_types application/json;
    gzip_min_length 1000;

    ssl_certificate /etc/letsencrypt/live/**<mydomain>**/fullchain.pem;
    ssl_certificate_key /etc/letsencrypt/live/**<mydomain>**/privkey.pem;

    location / {
        include uwsgi_params;
        uwsgi_pass unix:/mnt/sock/api.sock;
    }
}

```

**/mnt/sock/ folder is with 777 permissions.**

**example of api.ini file:**

```
[uwsgi]
plugins = python3
socket = /mnt/sock/api.sock
processes = 4
threads = 2

chdir = /root/virtualenv/bemani/wsgi
virtualenv = /root/virtualenv
wsgi-file = api.wsgi
callable = app
```

**example of services.ini file:**

```
[uwsgi]
plugins = python3
socket = /mnt/sock/services.sock
processes = 4
threads = 2

virtualenv = /root/virtualenv
chdir = /root/virtualenv/bemani/wsgi
wsgi-file = services.wsgi
callable = app

```

nginx error log:

2021/05/10 17:55:27 [crit] 26923#26923: *35 connect() to unix:/mnt/sock/frontend.sock failed (13: Permission denied) while connecting to upstream, client: **<myIP>**, server: **<mydomain>**, request: "GET / HTTP/1.1", upstream: "uwsgi://unix:/mnt/sock/frontend.sock:", host: "**<mydomain>**:8573"
2021/05/10 17:55:33 [crit] 26923#26923: *38 connect() to unix:/mnt/sock/api.sock failed (13: Permission denied) while connecting to upstream, client: **<myIP>**, server: "**<mydomain>**, request: "GET / HTTP/1.1", upstream: "uwsgi://unix:/mnt/sock/api.sock:", host: ""**<mydomain>**:9573"
2021/05/10 17:55:34 [crit] 26923#26923: *38 connect() to unix:/mnt/sock/api.sock failed (13: Permission denied) while connecting to upstream, client: **<myIP>**, server: "**<mydomain>**, request: "GET /favicon.ico HTTP/1.1", upstream: "uwsgi://unix:/mnt/sock/api.sock:", host: "**<mydomain>**:9573", referrer: "https://"**<mydomain>**:9573/"

Any help would be much appreciated ^^ I wonder if there was something I missed..

Basically im using normal Chrome browser to check the services for any output. I was mostly hoping to see something on the front end such as sign in screen, like there was with the development tools.

**Comment by Ikfes** [comment]
Solved the issue by setting permissions for the .sock files.

I did the worst and just used chmod 777 for each of them.

It does seem to work now. ^^

**Comment by Ikfes** [comment]
Services do work, but I seem to get network issue when attempting to launch the client.

The config on the client is configured like this:

```
<network>
    <timeout __type="u32">30000</timeout>
    <sz_xrpc_buf __type="u32">102400</sz_xrpc_buf>
    <ssl __type="bool">0</ssl>
    <services __type="str">http://<MyDomain>:573/</services>
</network>
```

Here are logs.

```

[pid: 26934|app: 0|req: 1/1] 139.28.184.2 () {56 vars in 1021 bytes} [Mon May 10 18:20:08 2021] GET / => generated 33 bytes in 5 msecs (HTTP/1.1 401) 3 headers in 147 bytes (1 switches on core 0)
[pid: 26940|app: 0|req: 1/1] 139.28.184.2 () {42 vars in 770 bytes} [Mon May 10 18:20:01 2021] GET / => generated 249 bytes in 2 msecs (HTTP/1.1 308) 3 headers in 129 bytes (1 switches on core 0)
[pid: 26940|app: 0|req: 2/2] <MyIP> () {40 vars in 641 bytes} [Mon May 10 18:37:19 2021] GET / => generated 249 bytes in 1 msecs (HTTP/1.1 308) 3 headers in 129 bytes (1 switches on core 1)
Received request:
<call model="KFC:J:A:A:2019020600" srcid="<MyBCPID>" tag="3b3b0773">
    <services method="get">
        <info>
            <AVS2 __type="str">2.15.8 r6631</AVS2>
        </info>
    </services>
</call>

Sending response:
<response dstid="<MyBCPID>">
    <services expire="600" mode="operation" product_domain="1" status="0">
        <item name="cardmng" url="http://<ServerIP>:80/" />
        <item name="dlstatus" url="http://<ServerIP>:80/" />
        <item name="eacoin" url="http://<ServerIP>:80/" />
        <item name="facility" url="http://<ServerIP>:80/" />
        <item name="lobby" url="http://<ServerIP>:80/" />
        <item name="local" url="http://<ServerIP>:80/" />
        <item name="message" url="http://<ServerIP>:80/" />
        <item name="package" url="http://<ServerIP>:80/" />
        <item name="pcbevent" url="http://<ServerIP>:80/" />
        <item name="pcbtracker" url="http://<ServerIP>:80/" />
        <item name="pkglist" url="http://<ServerIP>:80/" />
        <item name="posevent" url="http://<ServerIP>:80/" />
        <item name="local2" url="http://<ServerIP>:80/" />
        <item name="ntp" url="ntp://pool.ntp.org/" />
        <item name="keepalive" url="http://<ServerIP>/core/keepalive?pa=<ServerIP>&amp;ia=<ServerIP>&amp;ga=<ServerIP>&amp;ma=<ServerIP>&amp;t1=2&amp;t2=10" />
    </services>
</response>

```
```
[pid: 26949|app: 0|req: 1/3] <MyIP> () {36 vars in 524 bytes} [Mon May 10 18:45:36 2021] POST /?model=KFC:J:A:A:2019020600&module=services&method=get => generated 1084 bytes in 49 msecs (HTTP/1.0 200) 4 headers in 131 bytes (1 switches on core 0)
Received request:
<call model="KFC:J:A:A:2019020600" srcid="<MyBCPID>" tag="02d2350b">
    <services method="get">
        <info>
            <AVS2 __type="str">2.15.8 r6631</AVS2>
        </info>
    </services>
</call>

Sending response:
<response dstid="<MyBCPID>">
    <services expire="600" mode="operation" product_domain="1" status="0">
        <item name="cardmng" url="http://<ServerIP>:80/" />
        <item name="dlstatus" url="http://<ServerIP>:80/" />
        <item name="eacoin" url="http://<ServerIP>:80/" />
        <item name="facility" url="http://<ServerIP>:80/" />
        <item name="lobby" url="http://<ServerIP>:80/" />
        <item name="local" url="http://<ServerIP>:80/" />
        <item name="message" url="http://<ServerIP>:80/" />
        <item name="package" url="http://<ServerIP>:80/" />
        <item name="pcbevent" url="http://<ServerIP>:80/" />
        <item name="pcbtracker" url="http://<ServerIP>:80/" />
        <item name="pkglist" url="http://<ServerIP>:80/" />
        <item name="posevent" url="http://<ServerIP>:80/" />
        <item name="local2" url="http://<ServerIP>:80/" />
        <item name="ntp" url="ntp://pool.ntp.org/" />
        <item name="keepalive" url="http://<ServerIP>/core/keepalive?pa=<ServerIP>&amp;ia=<ServerIP>&amp;ga=<ServerIP>&amp;ma=<ServerIP>&amp;t1=2&amp;t2=10" />
    </services>
</response>
```

```
[pid: 26949|app: 0|req: 2/4] <MyIP> () {36 vars in 524 bytes} [Mon May 10 18:48:31 2021] POST /?model=KFC:J:A:A:2019020600&module=services&method=get => generated 1084 bytes in 11 msecs (HTTP/1.0 200) 4 headers in 131 bytes (1 switches on core 1)
Received request:
<call model="KFC:J:A:A:2019020600" srcid="<MyBCPID>" tag="2ed622cc">
    <services method="get">
        <info>
            <AVS2 __type="str">2.15.8 r6631</AVS2>
        </info>
    </services>
</call>

Sending response:
<response dstid="<MyBCPID>">
    <services expire="600" mode="operation" product_domain="1" status="0">
        <item name="cardmng" url="http://<ServerIP>:80/" />
        <item name="dlstatus" url="http://<ServerIP>:80/" />
        <item name="eacoin" url="http://<ServerIP>:80/" />
        <item name="facility" url="http://<ServerIP>:80/" />
        <item name="lobby" url="http://<ServerIP>:80/" />
        <item name="local" url="http://<ServerIP>:80/" />
        <item name="message" url="http://<ServerIP>:80/" />
        <item name="package" url="http://<ServerIP>:80/" />
        <item name="pcbevent" url="http://<ServerIP>:80/" />
        <item name="pcbtracker" url="http://<ServerIP>:80/" />
        <item name="pkglist" url="http://<ServerIP>:80/" />
        <item name="posevent" url="http://<ServerIP>:80/" />
        <item name="local2" url="http://<ServerIP>:80/" />
        <item name="ntp" url="ntp://pool.ntp.org/" />
        <item name="keepalive" url="http://<ServerIP>/core/keepalive?pa=<ServerIP>&amp;ia=<ServerIP>&amp;ga=<ServerIP>&amp;ma=<ServerIP>&amp;t1=2&amp;t2=10" />
    </services>
</response>

[pid: 26949|app: 0|req: 3/5] <MyIP> () {36 vars in 524 bytes} [Mon May 10 18:50:29 2021] POST /?model=KFC:J:A:A:2019020600&module=services&method=get => generated 1084 bytes in 11 msecs (HTTP/1.0 200) 4 headers in 131 bytes (1 switches on core 0)

```

**Comment by Ikfes** [comment]
I did fix the port in server.yaml to match port 573

The error code changed though.

5-2002-2502 right now.

**EDIT: I changed the default nginx server bracket from port 80 to 8080 just to be sure.**

**I just wonder if the server.yaml port should match the nginx configuration port for the services. In other words, should I change the server.yaml port to 573 as well?**

**EDIT2: That did the trick. ^^**

**EDIT3:** Well shit, I ended up making almost a complete tutorial by accident when backtracking my own mistakes. If the author wants, he can delete this issue/post or message and I can also delete it myself, or let it remain here.

**Comment by DragonMinded** [comment]
Haha yeah, the way services works is confusing and unfortunately means there needs to be a config for the port which matches your outermost connection point (nginx in this case). I touch on that in https://github.com/dragonminded/bemaniutils#logs-show-that-games-only-request-the-initial-services-packet-additional-packets-are-not-sent-and-games-do-not-go-online in the readme. If you see a way to phrase it better or if you want to add in a section in the productionizing portion about making sure the socket has the right permissions for both nginx and uwsgi (usually handled by changing user/group permissions and making sure both are either in the same group, or that the file is owned by one and group owned by another), go for it!

**Comment by Ikfes** [comment]
Another thing was modulenotfound error (it was looking for bemani) when running the uwsgi command so I did the thing what any certified idiot would do and copied the entire bemani folder from above the wsgi again into bemani/wsgi folder and the app ran just fine haha. I wonder if there was a better way? :P
